src/Agent.Worker/TestResults/TestRunPublisher.cs

```diff
@@ -0,0 +1,311 @@
+        {
+            _executionContext = executionContext;
+            _projectName = projectName;
+            _runContext = runContext;
+            _resultReader = resultReader;
+            connection.InnerHandler.Settings.SendTimeout = TimeSpan.FromSeconds(PUBLISH_TIMEOUT);
+            _httpClient = connection.GetClient<TestManagementHttpClient>();
+            _zipper = HostContext.GetService<IZipper>();
+        }
+
+        public void InitializePublisher(IExecutionContext executionContext, ITestManagementHttpClient httpClient, string projectName, TestRunContext runContext, IResultReader resultReader)
+        {
+            _executionContext = executionContext;
+            _projectName = projectName;
+            _runContext = runContext;
+            _resultReader = resultReader;
+            _httpClient = httpClient;
+            _zipper = HostContext.GetService<IZipper>();
+        }
+
+        /// <summary>
+        /// Publishes the given results to the test run.
+        /// </summary>
+        /// <param name="testResults">Results to be published.</param>
+        public void AddResults(TestCaseResultData[] testResults)
+        {
+            int noOfResultsToBePublished = BATCH_SIZE;
+
+            for (int i = 0; i < testResults.Length; i += BATCH_SIZE)
+            {
+                if (i + BATCH_SIZE >= testResults.Length)
+                {
+                    noOfResultsToBePublished = testResults.Length - i;
+                }
+                _executionContext.Debug($"Test results remaining: {(testResults.Length - i)}");
+
+                var currentBatch = new TestCaseResultData[noOfResultsToBePublished];
+                Array.Copy(testResults, i, currentBatch, 0, noOfResultsToBePublished);
+
+                Task<List<TestCaseResult>> testresults = _httpClient.AddTestResultsToTestRunAsync(currentBatch, _projectName, _testRun.Id);
+
+                testresults.Wait();
+
+                for (int j = 0; j < noOfResultsToBePublished; j++)
+                {
+                    // Do not upload duplicate entries 
+                    string[] attachments = testResults[i + j].Attachments;
+                    if (attachments != null)
+                    {
+                        Hashtable attachedFiles = new Hashtable(StringComparer.CurrentCultureIgnoreCase);
+                        foreach (string attachment in attachments)
+                        {
+                            if (!attachedFiles.ContainsKey(attachment))
+                            {
+                                TestAttachmentRequestModel reqModel = GetAttachmentRequestModel(attachment);
+                                if (reqModel != null)
+                                {
+                                    Task<TestAttachmentReference> trTask = _httpClient.CreateTestResultAttachmentAsync(reqModel, _projectName, _testRun.Id, testresults.Result[j].Id);
```

Comment:
we are not using Wait() in the entire core agent. everything is async/await pattern.